src/cpp/client/wrapped_credentials.cc

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+#include "src/cpp/client/wrapped_credentials.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <grpc/grpc.h>
+#include <grpc/grpc_security.h>
+#include <grpcpp/security/credentials.h>
+#include <grpcpp/support/channel_arguments.h>
+
+#include "src/core/lib/iomgr/exec_ctx.h"
+#include "src/core/lib/security/credentials/credentials.h"
+
+namespace grpc {
+
+class WrappedChannelCredentials final : public ChannelCredentials {
```

Comment:
I don't understand the conflict here.  First, I didn't actually propose removing the check for `c_creds()` being null for channel creds, only for call creds.  And second, even if I had, I don't see why that would prevent us from removing this wrapper class.

I think this should be structured the same way as the server creds, except that we don't need the equivalent of the `SecureServerCredentials` class, because there is no `AuthMetadataProcessor` issue on the client side.  In other words:
- The base class has the data member, a protected ctor that takes the c_creds arg, and a protected  non-virtual accessor method.
- For creds subclasses that are backed by C-core creds types, they can pass the C-core object to the base class ctor.
- For creds subclasses not backed by C-core creds types, they can pass nullptr to the base class ctor.

Reply:
> I didn't actually propose removing the check for c_creds() being null for channel creds

That's the only point of contention. I made some of your proposed changes. We are left with a very minimal `WrappedChannelCredentials` type that serves to hide the public constructor. It would be identical code for ~10 different credentials types if we wanted to remove the wrapper. We can if you think having different concrete types is meaningful polymorphism.